### PR TITLE
XCMP, xTokens, Assets, AssetManager and Treasury for Manta 

### PR DESCRIPTION
**Pull request by ghzlatarev**
## Description

Adds the following functionalities on Manta parachain, in order to start connecting to sibling chains and suporting their assets:

* XCMP queue for parachain to parachain communication and related XCM configurations
* Assets for supporting foreign chain assets
* Asset Manager to manage the aforementioned assets
* xTokens for cross-chain assets transfers

* Weights of new pallets benchmarked with https://github.com/Manta-Network/Manta/actions/runs/4545375784
* Moved some test only code to runtime/common and reusing it between all runtimes

---

Before we can approve this PR for merge, please make sure that **all** the following items have been checked off:
- [x] Connected to an issue with discussion and accepted design using zenhub "Connect issue" button below
- [x] Added **one** label out of the `L-` group to this PR
- [x] Added **one or more** labels from the `A-` and `C-` groups to this PR
- [x] Explicitly labelled `A-calamari`, `A-dolphin` and/or `A-manta` if your changes are meant for/impact either of these (CI depends on it)
- [x] Re-reviewed `Files changed` in the Github PR explorer.

Situational Notes:
- If adding functionality, write unit tests!
- If importing a new pallet, choose a proper module index for it, and allow it in `BaseFilter`. Ensure **every** extrinsic works from front-end. If there's corresponding tool, ensure both work for each other.
- If needed, update our Javascript/Typescript APIs. These APIs are officially used by exchanges or community developers.
- If modifying existing runtime storage items, make sure to implement storage migrations for the runtime and test them with `try-runtime`. This includes migrations inherited from upstream changes, and you can search the diffs for modifications of `#[pallet::storage]` items to check for any.

### CODE REVIEW

File: runtime/manta/src/lib.rs [hunk 1]
```diff
@@ -199,6 +203,8 @@ impl Contains<Call> for MantaFilter {
             | Call::Multisig(_)
             | Call::AuthorInherent(pallet_author_inherent::Call::kick_off_authorship_validation {..}) // executes unsigned on every block
             | Call::Balances(_)
+            | Call::XTokens(orml_xtokens::Call::transfer {..}
+                | orml_xtokens::Call::transfer_multicurrencies {..})
```
Comment (ghzlatarev) [inline runtime/manta/src/lib.rs]: todo: dont need this

File: runtime/manta/src/lib.rs [hunk 2]
```diff
@@ -479,6 +469,36 @@ impl manta_collator_selection::Config for Runtime {
     type CanAuthor = AuraAuthorFilter;
 }
 
+parameter_types! {
+    pub const ProposalBond: Permill = Permill::from_percent(1);
+    pub const ProposalBondMinimum: Balance = 500 * MANTA;
+    pub const ProposalBondMaximum: Balance = 10_000 * MANTA;
+    pub SpendPeriod: BlockNumber = 6 * DAYS;
+    pub const Burn: Permill = Permill::from_percent(0);
+    pub const TreasuryPalletId: PalletId = TREASURY_PALLET_ID;
```
Comment (ghzlatarev) [inline runtime/manta/src/lib.rs]: todo: think about these numebrs

File: runtime/manta/src/weights/cumulus_pallet_xcmp_queue.rs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+// --execution=wasm
+// --wasm-execution=compiled
+// --heap-pages=4096
+// --output=./scripts/benchmarking/frame-weights-output/cumulus_pallet_xcmp_queue.rs
+// --template=.github/resources/frame-weight-template.hbs
+
+#![cfg_attr(rustfmt, rustfmt_skip)]
+#![allow(unused_parens)]
+#![allow(unused_imports)]
+#![allow(clippy::unnecessary_cast)]
```
Comment (ghzlatarev) [inline runtime/manta/src/weights/cumulus_pallet_xcmp_queue.rs:41]: todo: re-benchmark for manta

File: runtime/manta/tests/common.rs [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+// along with Manta.  If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg(test)]
+
+pub mod xcm_mock;
+
+use sp_std::vec;
+use xcm::{
+    latest::{
+        prelude::{
```
Comment (ghzlatarev) [inline runtime/manta/tests/common.rs]: todo: could be deduplicated

File: runtime/manta/src/xcm_config.rs [hunk 5]
```diff
@@ -14,33 +14,103 @@
...
+    type ReservedDmpWeight = ReservedDmpWeight;
+    type OutboundXcmpMessageSource = XcmpQueue;
+    type XcmpMessageHandler = XcmpQueue;
+    type ReservedXcmpWeight = ReservedXcmpWeight;
+    type OnSystemEvent = ();
+    type CheckAssociatedRelayNumber = RelayNumberStrictlyIncreases;
+}
+
+parameter_types! {
+    pub const RelayNetwork: NetworkId = NetworkId::Kusama;
```
Comment (zqhxuyuan) [inline runtime/manta/src/xcm_config.rs]: should be Polkadot

File: runtime/manta/src/xcm_config.rs [hunk 6]
```diff
@@ -79,23 +166,56 @@ pub type Barrier = (
...
+    XcmFeesAccount,
+>;
+
 pub struct XcmExecutorConfig;
 impl Config for XcmExecutorConfig {
     type Call = Call;
     type XcmSender = XcmRouter;
     // Defines how to Withdraw and Deposit instruction work
-    type AssetTransactor = (); // MultiAssetTransactor;
+    type AssetTransactor = MultiAssetTransactor; // MultiAssetTransactor;
```
Comment (zqhxuyuan) [inline runtime/manta/src/xcm_config.rs]: remove the comment?

File: runtime/manta/src/assets_config.rs [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+parameter_types! {
+    pub const StartNonNativeAssetId: MantaAssetId = 8;
+    pub const NativeAssetId: MantaAssetId = 1;
+    pub NativeAssetLocation: AssetLocation = AssetLocation(
+        VersionedMultiLocation::V1(SelfReserve::get()));
+    pub NativeAssetMetadata: AssetRegistryMetadata<Balance> = AssetRegistryMetadata {
+        metadata: AssetStorageMetadata {
+            name: b"Manta".to_vec(),
+            symbol: b"KMA".to_vec(),
```
Comment (zqhxuyuan) [inline runtime/manta/src/assets_config.rs]: Symbol on Manta is still KMA?

Comment (ghzlatarev) [reply runtime/manta/src/assets_config.rs]: oops thanks for the catch.

File: runtime/manta/src/lib.rs [hunk 8]
```diff
@@ -754,6 +789,7 @@ impl_runtime_apis! {
 
     #[cfg(feature = "runtime-benchmarks")]
     impl frame_benchmarking::Benchmark<Block> for Runtime {
+
```
Comment (ferrell-code) [inline runtime/manta/src/lib.rs]: nit: could delete extra line

File: runtime/manta/tests/xcm_mock/mod.rs [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    }
+}
+
+decl_test_relay_chain! {
+    pub struct Relay {
+        Runtime = relay_chain::Runtime,
+        XcmConfig = relay_chain::XcmExecutorConfig,
+        new_ext = relay_ext(),
+    }
+}
```
Comment (ferrell-code) [inline runtime/manta/tests/xcm_mock/mod.rs:66]: Could use real polkadot runtime, applies to Calamari runtime too. Like here: https://github.com/AcalaNetwork/Acala/blob/54feaa65d729ea1e4f269d52fd4cbb2d0e304a48/runtime/integration-tests/src/relaychain/kusama_test_net.rs#L31

Could be a followup issue to this PR as well

File: runtime/manta/src/lib.rs [hunk 10]
```diff
@@ -65,9 +66,12 @@ use runtime_common::{
 };
 use session_key_primitives::{AuraId, NimbusId, VrfId};
 
+use xcm::latest::prelude::*;
```
Comment (ferrell-code) [inline runtime/manta/src/lib.rs:70]: Not sure if this should pass CI, but I get compiler warning of unused import here

File: runtime/manta/tests/xcm_mock/mod.rs (hunk 9 above)
Comment (ghzlatarev) [reply runtime/manta/tests/xcm_mock/mod.rs:66]: Issue already exists https://github.com/Manta-Network/Manta/issues/648

**Review by ferrell-code — DISMISSED** [review]

File: runtime/manta/src/lib.rs [hunk 11]
```diff
@@ -773,14 +809,102 @@ impl_runtime_apis! {
...
+            parameter_types! {
+                pub const TrustedTeleporter: Option<(MultiLocation, MultiAsset)> = None;
+                pub const TrustedReserve: Option<(MultiLocation, MultiAsset)> = Some((
+                    DotLocation::get(),
+                    // Random amount for the benchmark.
+                    MultiAsset { fun: Fungible(1_000_000_000_000), id: Concrete(DotLocation::get()) },
+                ));
+                pub const CheckedAccount: Option<AccountId> = None;
+                pub const DotLocation: MultiLocation = MultiLocation::parent();
+                pub MantaLocation: MultiLocation = MultiLocation::new(1, X1(Parachain(2104)));
```
Comment (Garandor) [inline runtime/manta/src/lib.rs]: use `MANTA_PARACHAIN_ID` define
if this is also hardcoded in calamari, pls change  too

File: runtime/manta/src/lib.rs [hunk 12]
```diff
@@ -479,6 +469,36 @@ impl manta_collator_selection::Config for Runtime {
     type CanAuthor = AuraAuthorFilter;
 }
 
+parameter_types! {
+    pub const ProposalBond: Permill = Permill::from_percent(1);
+    pub const ProposalBondMinimum: Balance = 3 * MANTA;
+    pub const ProposalBondMaximum: Balance = 50 * MANTA;
+    pub SpendPeriod: BlockNumber = 6 * DAYS;
```
Comment (Garandor) [inline runtime/manta/src/lib.rs]: IIRC we'd want manta to be a slower calamari, not sure what the correct factor is, but i'd consider bumping this to 14 days ( similar for voting )

Comment (ghzlatarev) [reply runtime/manta/src/lib.rs]: Makes sense.

**Review by zqhxuyuan — APPROVED** [review]

**Review by Garandor — APPROVED** [review]

**Pull request merged**